### PR TITLE
[HOLD] Don't use CREATED action as last message for LHN

### PR DESCRIPTION
**Pull request by aldo-expensify**
Fixes regression from: https://github.com/Expensify/App/pull/20223

When the reported action is the only action, we ended up returning the text of the report action with type CREATED. Normally we ignore this action and the new code wasn't ignoring it.

### Fixed Issues

$ https://github.com/Expensify/App/issues/20940
PROPOSAL: GH_LINK_ISSUE(COMMENT)

### Tests "__FAKE__"

1. Create a chat between two accounts
2. Send a SINGLE message from account A to account B
3. With account B, flag the message as harassment
4. Verify that the LHN shows "This is the beginning of your chat with concierge" and not "__FAKE__"

### Tests - last report message incorrect without opening the report

1. Flag the last message of a conversation
2. Go to another report
3. Log out / log in
4. Don't open the report where you flagged the message!
5. Verify that the flagged message doesn't appear as the last message of the report

### Tests - Group

1. Create a group with accounts A and B
2. Send a message with account A
3. Send a message with account B
4. With account A, flag the message sent by account B
5. Verify that the last message shown in the LHN shows the correct author (account A) and the correct message text.

### Tests - Task

1. Create a task for account B using account A
2. Send a message with account B
3. Flag the message with account A
4. Verify that the LHN last message is correct

### Tests - Offline

1. In a chat with account A and account B, send a message with account A
2. Go offline
3. Using account B, flag the message as harassment
4. Check that the optimistic last message in the LHN is correct
5. Go online and check that everything is fine

### Tests - error (only if you can manipulate the PHP API)

1. Add a throw at the start of the `try` block ([here](https://github.com/Expensify/Web-Expensify/blob/6375d056e546b72a4f8ec9c6883bf4c3bc23c890/lib/ReportAPI.php#L886))
1. In a chat with account A and account B, send a message with account A
3. Using account B, flag the message as harassment
4. Check that there is an error and that the LHN last message is reverted to the message we were not able to flag

- [x] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Review by neil-marcellini — DISMISSED** [review]
Thanks for fixing my mistake!

**Comment by aldo-expensify** [comment]
@neil-marcellini no problem! I removed some unnecessary `|| {}` from the code, can you re-approve? 🙏

**Comment by aldo-expensify** [comment]
@eVoloshchak are you online now to review this?

**Review by thienlnam — DISMISSED** [review]

**Comment by aldo-expensify** [comment]
> @aldo-expensify The LHN does not show `No activity yet` instantly. I think it shows that only when it receives the data from pusher. Is that expected?

That sounds like a good guess to me, maybe we need better optimistic data to show it instantly. Having said that, I would consider that unrelated and leave it for a follow up if we want to improve it.

cc @neil-marcellini

**Comment by aldo-expensify** [comment]
> Also, it does not seem to work if you're offline.
> 
>  Screen.Recording.2023-06-17.at.4.28.05.AM.mov

Yeah, I think that points to the same think mentioned in your other comment, we are probably not  flagging the comment optimistically and we wait to receive it in the response or in the pusher. I think that can be improved, but that is not related to the deploy blocker.

**Comment by aldo-expensify** [comment]
Good thinking in testing with threads, I didn't think about that.

**Comment by aldo-expensify** [comment]
We will have to check what is the current behaviour for those cases in production. If it is the same, then https://github.com/Expensify/App/pull/20223 didn't break it and we would be fine with proceeding. If production works differently, maybe then we should just revert https://github.com/Expensify/App/pull/20223?

**Comment by aldo-expensify** [comment]
Thanks for testing in production.

@neil-marcellini @thienlnam @dangrous what are you thoughts here? Should we revert https://github.com/Expensify/App/pull/20223 ? or do you see an easy fix?

**Comment by neil-marcellini** [comment]
Thanks for your testing @allroundexperts! It sounds like we have a handful of issues.
1. Comments are not [flagged pendingRemove optimistically](https://github.com/Expensify/App/pull/20949#issuecomment-1595473095) - Woops I think I missed this, should be an easy fix as @aldo-expensify mentioned
2. Same thing when offline. I thought I tested that, but maybe not 🤔 
3. First [flagged message on a task doesn't hide](https://github.com/Expensify/App/pull/20949#issuecomment-1595483433) - Very weird we should investigate.
4. ["Concierge: " is displayed in the LHN](https://github.com/Expensify/App/pull/20949#issuecomment-1595486497) after flagging a group chat comment. Also very strange, let's investigate.

Personally I think we should not revert the PR and fix these bugs one at a time since they are edge cases. I think we could easily fix the optimistic data with this PR. These don't feel like critical blockers, more like bugs related to a new feature. We should definitely still fix the issues eventually.

What do you think @aldo-expensify? Do you want to tackle this or split up the work with me?

**Comment by aldo-expensify** [comment]
> split up the work with me?

If you know where to do these changes it would save me some time, so I would appreciate it you take part of the work.

**Comment by aldo-expensify** [comment]
Pushed a fix for the group case: https://github.com/Expensify/App/pull/20949#issuecomment-1595486497

I'm going to work on the tasks case now.

**Comment by aldo-expensify** [comment]
@allroundexperts or maybe wait a sec, I think there will be another bug:

- Create a group with accounts A and B
- Send a message with account A
- Send a message with account B
- With account A, flag the message sent by account B

I'm pretty sure that the last message will show account A's first message with B being the author

👁️

**Comment by aldo-expensify** [comment]
This will also be wrong if there was an attachment before the flagged message:
 https://github.com/Expensify/App/blob/24b8f5a99f182da68e455c2d5a89c6924861f783/src/libs/OptionsListUtils.js#L378-L379

**Comment by aldo-expensify** [comment]
Another case that doesn't work:

1. Flag the last message of a conversation
2. Go to another report
3. Log out / log in
4. Don't open the report where you flagged the message!
5. Verify that the flagged message appears as the last message of the report and it will disappear if you open the report

File: src/pages/home/sidebar/SidebarLinks.js
Comment (aldo-expensify) [inline src/pages/home/sidebar/SidebarLinks.js:301]: can we wrap this `lodashGet(reportAction, 'message[0].moderationDecisions[0].decision')` in a `ReportActionUtils` function?

File: src/libs/actions/Report.js
Comment (aldo-expensify) [inline src/libs/actions/Report.js:883]: We were not reverting the changes to the report if the `DeleteComment` command failed

File: src/libs/actions/Report.js
Comment (aldo-expensify) [inline src/libs/actions/Report.js:804]: Put this code into a function to use it for the `FlagComment` and `DeleteComment` commands.

File: src/libs/actions/Report.js
Comment (aldo-expensify) [reply src/libs/actions/Report.js:883]: Before 🐛 :



https://github.com/Expensify/App/assets/87341702/6a0eae95-6a02-4c8f-90dd-c246eab05287


After [FIXED]:

https://github.com/Expensify/App/assets/87341702/cfb8f9a0-264c-44a3-9669-96fce225e6b1

**Comment by aldo-expensify** [comment]
This PR is ready for review again.

@allroundexperts there is a backend PR that is in review that should fix what the backend sends as "lastMessageText", etc, when you flag a message. Basically it filters out the report action like the front end is doing. That PR not being deployed yet may make it more difficult to test stuff.

File: src/libs/actions/Report.js
Comment (luacmartins) [inline src/libs/actions/Report.js:804]: If we keep thinking I'm sure we'll come up with a longer name 🤣 Can we try to shorten it to `getLastVisibleMessageOnyxUpdatesAfterActionChange`?

**Review by luacmartins — COMMENTED** [review]
Left a couple of comments

File: src/libs/ReportUtils.js
Comment (aldo-expensify) [reply src/libs/ReportUtils.js:1004]: In fact, that check doesn't make sense... I see that above we are already accessing `report.lastMessageText` (line 1001) in an unsafe way

Comment (aldo-expensify) [reply src/libs/ReportUtils.js:1004]: The original code, previous to Neil changes, already had this check for `report ?` that didn't make sense:

image

I think it is safe to assume that the report is ALWAYS going to be there

File: src/libs/actions/Report.js
Comment (aldo-expensify) [reply src/libs/actions/Report.js]: Thanks for leaving a suggested name, I replaced it.

File: src/libs/ReportUtils.js
Comment (aldo-expensify) [inline src/libs/ReportUtils.js:1004]: @luacmartins is this ok? or still prefer lodashget?

File: src/libs/ReportUtils.js
Comment (luacmartins) [reply src/libs/ReportUtils.js]: sounds good

File: src/libs/ReportUtils.js
Comment (luacmartins) [reply src/libs/ReportUtils.js:1004]: that's fine

**Review by luacmartins — DISMISSED** [review]

**Review by neil-marcellini — DISMISSED** [review]
**Tests "FAKE"**
Not passing for me. The last message shows a whisper. I pulled the latest from Auth, and rebuilt and restarted it before testing.

https://github.com/Expensify/App/assets/26260477/c9ed5e89-0bf0-46ab-aca9-b71f87e20009



Unrelated bug, probably on main, maybe dev only?
1. Sign in with an account A and create a chat with a new user `example@expensifail.com`
2. On another device attempt to login as that user

Expected result: The user is prompted for the magic code
Actual result: An error is displayed `Cannot get account details, please try again or contact concierge@expensify.com`

**Comment by aldo-expensify** [comment]
> Unrelated bug, probably on main, maybe dev only?
> 
> 1. Sign in with an account A and create a chat with a new user `example@expensifail.com`
> 2. On another device attempt to login as that user
> 
> Expected result: The user is prompted for the magic code Actual result: An error is displayed `Cannot get account details, please try again or contact concierge@expensify.com`

Yes, unrelated. I think this is the same I reported here: https://expensify.slack.com/archives/C049HHMV9SM/p1686949198475289

**Comment by aldo-expensify** [comment]
> Tests "FAKE"
Not passing for me. The last message shows a whisper. I pulled the latest from Auth, and rebuilt and restarted it before testing.

Oh, I think I saw once this one... My suspicion is that there is an onyx update we push with the concierge message as the last message for the report. I'm guessing that needs a fix in Web-E. I didn't see this case much because my pusher events sometimes don't work in dev.

I don't think we should hold this App PR on that as that involves a fix in Web-E.

**Comment by neil-marcellini** [comment]
Since I made commits on this PR it would be best for me not to review.

**Comment by luacmartins** [comment]
This is no longer a blocker, so removing the CP staging label.

**Comment by aldo-expensify** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

**Review by thienlnam — COMMENTED** [review]
@allroundexperts Are you able to run through the checklist for this?

**Comment by thienlnam** [comment]
Ah gotcha, nvm then - could you please put hold in the title then until we get the backend PR merged?

**Comment by aldo-expensify** [comment]
Sorry, not yet, the PR was not reviewed for a long time, ran into conflicts and now some tests are failing

**Comment by aldo-expensify** [comment]
https://github.com/Expensify/Web-Expensify/pull/37910 is in staging, I'll start working on resolving conflicts + retesting.

**Comment by aldo-expensify** [comment]
Worked on the conflicts, but I have to retest everything. Resolving conflicts I saw that some changes I was doing here were already done in `main`, so I'll go through the list of cases we had and check what is the state for them.

**Comment by aldo-expensify** [comment]
I'm going to close this, the original issue seems to be fixed and this is full of conflicts. If there is something to do, we should do it in a fresh new PR.

cc @neil-marcellini since I think you needed some changes from here, but maybe you already implemented them

**Pull request closed without merging**
